File: LeetCode/Medium/2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.cpp
class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {
        int n = questions.size();
        vector<long long> dp(n+1);

        for (int i = n - 1; i > -1; i--){
            int point = questions[i][0];
            int bp = questions[i][1];

            dp[i] = max(dp[i+1], dp[min(n, bp + i + 1)] + point);
        }

        return dp[0];
    }
};

File: LeetCode/Medium/2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    public long mostPoints(int[][] questions) {
        int n = questions.length;
        long[] dp = new long[n+1];

        for (int i = n-1; i > -1; i--){
            int point = questions[i][0];
            int bp = questions[i][1];

            dp[i] = Math.max(dp[i+1], dp[Math.min(n, bp+i+1)] + point);
        }

        return dp[0];
    }
}